app/chat/components/ChatMessage.tsx

```
@@ -0,0 +1,149 @@
+                                    <span className="text-sm tracking-wide text-gray-200">Agent is processing...</span>
+                                </div>
+                            ) : (
+                                <MarkdownRenderer
+                                    content={message.content}
+                                />
+                            )}
+                            {/* Render tutorial content if present */}
+                            {tutorialMdx && (
+                                <div className="mt-4 p-3 bg-gray-700/30 rounded-lg">
+                                    <h4 className="text-sm font-semibold mb-2 text-gray-100">Step {currentStep} of {totalSteps}</h4>
+                                    <MarkdownRenderer content={useMemo(() => transformMdxWithSnippets(tutorialMdx, tutorialSnippets || []), [tutorialMdx, tutorialSnippets])} />
```

Biome (2.1.2)

[error] This hook is being called conditionally, but all hooks must be called in the exact same order in every component render.

For React to preserve state between calls, hooks needs to be called unconditionally and always in the same order.
See https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

(lint/correctness/useHookAtTopLevel)